linode/manager
# ci: [M3-6837] - Allow Cypress runs to be parallelized without reliance on third party services

## Description 📝
This experimentally introduces yet another handful of environment variables to manipulate the behavior of Cloud Manager's Cypress tests. Specifically, `CY_TEST_SPLIT_RUN`, `CY_TEST_SPLIT_RUN_INDEX`, and `CY_TEST_SPLIT_RUN_TOTAL` have been added.

The purpose of these environment variables is to facilitate basic test parallelization which does not rely on any 3rd party service (e.g. Cypress Cloud). This implementation will not be as performant as what we're used to, because Cypress Cloud's parallelization uses historic test result data to determine which runners run which tests to better optimize run length.

Parallelization is achieved by limiting each process's specs to a subset of the total. It is up to the user (or CI pipeline) to initiate each process and interpret the results.

## Usage 🛠️
Use `CY_TEST_SPLIT_RUN` to enable the feature. Use `CY_TEST_SPLIT_RUN_TOTAL` to specify the total number of runners for a given test run, and use `CY_TEST_SPLIT_RUN_INDEX` to specify the runner index to assign to each individual runner.

For example, if you wish to run your Cypress tests locally in parallel, split between 2 runners, you could execute the following:

```bash
CY_TEST_SPLIT_RUN=1 CY_TEST_SPLIT_RUN_TOTAL=2 CY_TEST_SPLIT_RUN_INDEX=1 yarn cy:run
CY_TEST_SPLIT_RUN=1 CY_TEST_SPLIT_RUN_TOTAL=2 CY_TEST_SPLIT_RUN_INDEX=2 yarn cy:run
```
The console output for each should reflect that the run has been split, and each process should be handling roughly half of the specs. (Keep in mind that the tests themselves will likely fail unless you also specify a separate `MANAGER_OAUTH` value for each process.)

## Major Changes 🔄
- Added `CY_TEST_SPLIT_RUN`, `CY_TEST_SPLIT_RUN_INDEX`, and `CY_TEST_SPLIT_RUN_TOTAL`.

## How to test 🧪
* Confirm that tests run as expected without any of these environment variables applied (we can rely on the automated run for this one)
* Confirm that if `CY_TEST_SPLIT_RUN` is enabled, `CY_TEST_SPLIT_RUN_TOTAL` and `CY_TEST_SPLIT_RUN_INDEX` must also be defined, and Cypress will fail if not
* Confirm that if split running is enabled, only a subset of specs are run for each process, and that there is no overlap (e.g. if one runner runs a spec, the other runners should not run that spec)
